lib/IntermediateRepresentation/common.h

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_INTERMEDIATEREPRESENTATION_COMMON_H_
+#define FORTRAN_INTERMEDIATEREPRESENTATION_COMMON_H_
+
+#include "../common/idioms.h"
+#include "../common/indirection.h"
+#include "../evaluate/expression.h"
+#include "../evaluate/type.h"
+#include "../evaluate/variable.h"
+#include "../parser/parse-tree.h"
+#include "../semantics/symbol.h"
+
+// Some useful, self-documenting macros for failure modes
+#define STRINGIFY(X) #X
+#define LINE2STRING(X) STRINGIFY(X)
+#define AT_HERE " at " __FILE__ "(" LINE2STRING(__LINE__) ")"
+#define DIE Fortran::common::die
+#define SEMANTICS_FAILED(STRING) DIE("semantics bug: " STRING AT_HERE)
+#define SEMANTICS_CHECK(CONDITION, STRING) \
+  if (CONDITION) { \
+  } else { \
+    DIE("semantics bug: " STRING AT_HERE); \
+  }
+#define WRONG_PATH() DIE("control should not reach here" AT_HERE)
+
+namespace Fortran::IntermediateRepresentation {
+class Statement;
+class BasicBlock;
+struct Program;
+struct GraphWriter;
+
+struct Attribute {
+  enum { IntentIn, IntentOut, IntentInOut } attribute;
```

Comment:
This `enum` might be a useful thing to have in `lib/common/fortran.h` as an `enum class Intent`.

Should `Value` also be in that list?

Reply:
Yes, maybe they should go there.

Would `Value` mean pass-by-value here?  Relatedly, FIR will have to be able to represent and distinguish between references/pointers and values (that are not references).  Also with BIND(C), it makes some sense that `Call` is an abstraction over calling conventions.

Reply:
`Value` would mean "dummy argument has the VALUE attribute", and unfortunately that's not the same thing as pass-by-value.  In Fortran it basically means "pass by value, if it fits in a register, otherwise pass by address and immediately make a local copy on entry".  The FIR probably needs to be capable of describing argument-passing methods in terms of implementation as well or instead of their Fortran dummy argument attributes so that arguments to C & assembly code in runtime libraries can be characterized.

Reply:
Thanks for the explanation. I agree that it probably makes a lot of sense to add VALUE here with its strong interactions with INTENT as well.